### PR TITLE
Fix #9326: Warranty Expires should be also sortable

### LINKED ISSUE #9326: Warranty Expires should be also sortable
image

Would like to see that we can use warranty expires with sorting. Make it easy to find from main view.
Running version 5.1.2 in Azure Docker.

### PR DESCRIPTION
**Pull request by zacharesmer**
# Description

This adds the ability to sort by the warranty expires field. I added a generated column to the database that still calculates the warranty expires date on the fly, but allows it to be accessed like a normal database column. That change led to some redundancy in other places, so I cleaned that up. 

Also I changed the spelling of Warrantee in the name of getExpiringWarantee to Warranty to be consistent with the other parts of the project.

Regarding compatibility: generated columns are supported in MySQL since 5.7.1 and MariaDB since 10.2. I'm not totally sure from the requirements in the documentation whether this will be a problem.

Fixes #9326, but also slightly affects #5905. If #5925 or #524 are ever addressed, this shouldn't hurt anything.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

I unfortunately couldn't figure out how to get the automated tests running, but I tested by hand with seeded data in Firefox and Chrome. I added a variety of warranty info (this isn't seeded by default) and made sure it sorted correctly. It works with null values and updates correctly when a purchase date or warranty month field is modified.

There is one existing test in tests/Unit/AssetTest.php for the old warranty_expires mutator that should still cover this.

I also don't have email alerts set up, but I put a a breakpoint in the getExpiringWarranty function and it successfully found the assets that were expiring within the threshold I set.

**Test Configuration**:
* PHP version: 8.1.2
* MySQL version: 8.0.31
* Webserver version: built in artisan serve with laravel 8.83.22  
* OS version Ubuntu 22.04

# Checklist:

- [ ] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [ ] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by snipe** [comment]
Hi there - thanks for this, and sorry for the delay. My only concern with this is that while we don't officially support other databases (pgSQL, sqlite) I do know that people use them, and I'm concerned about how `virtualAs()` behaves there.

File: database/migrations/2022_12_02_170921_add_warranty_expires_to_assets.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
+    {
+        Schema::table('assets', function (Blueprint $table) {
+            $table->date('warranty_expires')->virtualAs('DATE_ADD(`purchase_date`, INTERVAL `warranty_months` MONTH)')->nullable()->default(null);
```
Comment (snipe) [inline database/migrations/2022_12_02_170921_add_warranty_expires_to_assets.php:17]: I'd much rather we just store this as a date column that we populate based on the other information we have when saving, or perhaps even allow the user to override manually. Virtual columns are not supported by some of the other SQL database types that we have to unofficially support. If we want to sort on a date, we should just store it as a regular date and do the math for what that date should be in the code.

**Review by snipe — CHANGES_REQUESTED** [review]
I don't think we can take this with the virtual column. There are more agnostic ways to handle this, which while they require a bit more logic in the code, will be more flexible with different *SQL versions.

File: database/migrations/2022_12_02_170921_add_warranty_expires_to_assets.php (hunk 1 above)
Comment (zacharesmer) [reply database/migrations/2022_12_02_170921_add_warranty_expires_to_assets.php:17]: Thanks! I will work on doing it that way instead.

**Comment by snipe** [comment]
Closing this for housekeeping since we haven't heard back in a bit - feel free to re-open though! (Though it's possible we may end up de-norming that field, since some folks want to edit it directly. Sigh.)

**Pull request closed without merging**